src/ucp/proto/ucp_ep.c

```diff
@@ -46,38 +46,98 @@ static void ucp_ep_progress_pending(ucs_callback_t *self)
     UCS_ASYNC_UNBLOCK(&worker->async);
 }
 
-ucs_status_t ucp_ep_create(ucp_worker_h worker, ucp_address_t *address,
-                           ucp_ep_h *ep_p)
+ucs_status_t ucp_ep_new(ucp_worker_h worker, uint64_t dest_uuid,
+                        const char *message, ucp_ep_h *ep_p)
 {
-    ucs_status_t status;
     ucp_ep_h ep;
 
     ep = ucs_calloc(1, sizeof(*ep), "ucp ep");
     if (ep == NULL) {
-        status = UCS_ERR_NO_MEMORY;
-        goto err;
+        ucs_error("Failed to allocate ep");
+        return UCS_ERR_NO_MEMORY;
     }
 
     ep->worker               = worker;
+    ep->uct_ep               = NULL;
     ep->config.max_short_tag = SIZE_MAX;
     ep->config.max_short_put = SIZE_MAX;
     ep->config.max_bcopy_put = SIZE_MAX;
     ep->config.max_bcopy_get = SIZE_MAX;
-    ep->notify.func          = ucp_ep_progress_pending;
+    ep->dest_uuid            = dest_uuid;
+    ep->rsc_index            = -1;
+    ep->dst_pd_index         = -1;
     ep->state                = 0;
     ucs_queue_head_init(&ep->pending_q);
+    ep->notify.func          = ucp_ep_progress_pending;
+    ep->wireup.aux_ep        = NULL;
+    ep->wireup.next_ep       = NULL;
+
+    sglib_hashed_ucp_ep_t_add(worker->ep_hash, ep);
+
+    ucs_debug("created ep %p 0x%"PRIx64"->0x%"PRIx64" %s", ep, worker->uuid,
+              ep->dest_uuid, message);
+    *ep_p = ep;
+    return UCS_OK;
+}
+
+void ucp_ep_ready_to_send(ucp_ep_h ep)
+{
+    ucp_worker_h worker          = ep->worker;
+    uct_iface_attr_t *iface_attr = &worker->iface_attrs[ep->rsc_index];
+
+    ucs_debug("connected 0x%"PRIx64"->0x%"PRIx64, worker->uuid, ep->dest_uuid);
+
+    ep->state               |= UCP_EP_STATE_READY_TO_SEND;
+    ep->config.max_short_tag = iface_attr->cap.am.max_short - sizeof(uint64_t);
+    ep->config.max_short_put = iface_attr->cap.put.max_short;
+    ep->config.max_bcopy_put = iface_attr->cap.put.max_bcopy;
+    ep->config.max_bcopy_get = iface_attr->cap.get.max_bcopy;
+}
+
+void ucp_ep_destroy_uct_ep_safe(ucp_ep_h ep, uct_ep_h uct_ep)
+{
+    ucs_assert(uct_ep != NULL);
```

Comment:
Since it is not critical path we can replace the assert with in actual check and "ERROR_PARAM" code ?